cri: Implement runtime handler support

sboeuf opened the pull request:
Fixes #1751

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
@mrunalp @runcom PTAL, I'd like to get some input about the first draft of the implementation of the runtime handler before I go further.

runcom commented on lib/config.go:
```diff
@@ -131,6 +131,12 @@ type RuntimeConfig struct {
+	// Runtimes defines a list of OCI compatible runtimes. The runtime to
+	// use is picked based on the runtime_handler provided by the CRI. If
+	// no runtime_handler is provided, the runtime will be picked based on
+	// the level of trust of the workload.
```
Do we still want to keep the default workload trust level? I though we could now get rid of it since kube dictates about the runtime to use. I don't have a strong opinion though. The linked issue also suggests to deprecate the trust level

sboeuf replied:
I know the issue suggests to deprecate the trust level, but what user of k8s are still using the old format to describe their pod? I guess I still want to make sure we won't break use case for those users.
@egernst WDYT?

runcom replied:
we can deprecate the feature, keep it for 2+ releases or so and get rid of it

sboeuf replied:
Yeah I like that. Let's just deprecate it, and when it becomes __hopefully__ widely used (after two more releases as you suggested), we will remove it :)

runcom commented on oci/oci.go:
```diff
@@ -95,6 +102,12 @@ type Runtime struct {
 	ctrStopTimeout           int64
 }
 
+// RuntimeHandler represents each item of the "crio.runtime.runtimes" TOML
```
I'd rather use `crio.runtimes` directly in the config

sboeuf replied:
So we would add a RuntimesConfig everywhere? I'm asking because CRI-O config seems pretty well organized with sections `runtime`/`image`/`api`, and it feels that adding `crio.runtimes` would add a new category, which is related to `runtime`.

runcom replied:
I believe `runtime` is too generic to be a top level at this point cause now that we have `runtimes`, the other options should very well be `runtime_options`. But even in that case, it's not that good. I don't know :smile:

runcom replied:
nevermind this comment anymore

TomSweeneyRedHat commented on oci/oci.go:
```diff
@@ -116,6 +129,23 @@ func (r *Runtime) Name() string {
+	if c.runtimeHandler != "" {
+		runtimeHandler, exist := r.runtimes[c.runtimeHandler]
+		if !exist {
+			if c.runtimeHandler == "untrusted" && r.untrustedPath != "" {
```
nit: Is there a common variable that could be used over a bare string for untrusted?

sboeuf replied:
Yes, I'll add a `const` for that!

sboeuf commented:
@runcom @mrunalp PR updated!

runcom commented on cmd/crio/config.go:
```diff
@@ -107,8 +113,17 @@ runtime_untrusted_workload = "{{ .RuntimeUntrustedWorkload }}"
+#
+# DEPRECATED: The runtime handler should provide a key to the map of runtimes,
+#     avoiding the need to rely on the level of trust of the workload to choose
+#     an appropriate runtime.
```
let's clarify that we'll support this for 1.12, no-op maybe in 1.13 and remove it altogether in 1.14

sboeuf replied:
Yes good point, let me add this comment.

runcom replied:
so, I also think we should print a warning in validateConfig if this is used from now on, basically just saying what's written here so admins and users are aware of this

sboeuf replied:
Both comments addressed!

runcom commented on oci/oci.go:
```diff
@@ -116,6 +129,23 @@ func (r *Runtime) Name() string {
+		return runtimeHandler.RuntimePath
+	}
+
 	if !c.trusted {
```
is everything below in line with the new additions above? I can still see untrustdPath references but I guess any inconsistencies would have had been caught in `validateConfig` right?

sboeuf replied:
I have add the code above so that in case of a `runtime_handler` provided by CRI, this will supersede the previous logic that is now below.
But I wanted to keep the same logic below in case we don't get any runtime handler, this way we keep the support of the deprecated options.

Also, once we completely remove untrusted runtime and the trust level (by 1.14 as agreed), this code will need to be revisited with some simpler names like `defaultRuntimePath` instead of `trustedPath`, and also we'll simplify the way the runtime handler is processed, since there won't be any fallback.

runcom replied:
totally agree

tallclair commented on cmd/crio/config.go:
```diff
@@ -107,7 +115,22 @@ runtime_untrusted_workload = "{{ .RuntimeUntrustedWorkload }}"
+  # The "crio.runtime.runtimes" table defines a list of OCI compatible runtimes.
+  # The runtime to use is picked based on the runtime_handler provided by the CRI.
+  # If no runtime_handler is provided, the runtime will be picked based on the level
+  # of trust of the workload.
```
I would leave out this line, since you're marking default_workload_trust as deprecated.

My interpretation of the new approach is that to implement the equivalent of `default_workload_trust=untrusted` without using deprecated options, you would set `runtime=<untrusted runtime>` (line 87), and then create another runtime named "trusted" (or any other name). Does that make sense?

tallclair replied:
I agree with keeping the implementation fallback, I was just wondering if it should be omitted from the documentation. Then again, I think it's fine to leave it, and just remove that when the deprecated options are removed.

tallclair commented on oci/oci.go:
```diff
@@ -95,6 +102,12 @@ type Runtime struct {
+// config table.
+type RuntimeHandler struct {
+	RuntimePath string `toml:"runtime_path"`
+}
```
:+1:  for using a nested struct. I think this might come in handy at some point.

tallclair commented on server/sandbox_run_linux.go:
```diff
@@ -311,6 +311,7 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 	g.SetHostname(hostname)
 
 	trusted := s.trustedSandbox(req)
+	runtimeHandler := req.GetRuntimeHandler()
```
IMO the RunPodSandbox request should fail if the requested runtime isn't configured. This is an important part of "failing closed" so an untrusted workload can't accidentally be run unsandboxed.

runcom replied:
agree with that

sboeuf replied:
Yes I understand the expectation here, but this will imply the modification of a bunch of things in CRI-O as the current design always expect `func (r *Runtime) Path(c *Container) string` to return something.
@runcom @mrunalp please let me know what you think here, I don't want to introduce too many changes without agreement first.

tallclair commented on oci/oci.go:
```diff
@@ -116,6 +129,23 @@ func (r *Runtime) Name() string {
 func (r *Runtime) Path(c *Container) string {
+	if c.runtimeHandler != "" {
+		runtimeHandler, ok := r.runtimes[c.runtimeHandler]
+		if !ok {
```
As commented elsewhere, I think this should be an error, rather than falling back to the default.

sboeuf replied:
Same here:
>Yes I understand the expectation here, but this will imply the modification of a bunch of things in CRI-O as the current design always expect `func (r *Runtime) Path(c *Container) string` to return something.
@runcom @mrunalp please let me know what you think here, I don't want to introduce too many changes without agreement first.

runcom replied:
this just affects `Path` so I would just start returning an error here when it makes sense (like here). Long term solution is a complete refactor of all of this but for now let's have a stopgap and return an error

tallclair commented on oci/oci.go:
```diff
@@ -116,6 +129,23 @@ func (r *Runtime) Name() string {
+		}
+
+		if runtimeHandler.RuntimePath == "" {
+			return r.trustedPath
```
how come this defaults to the trusted path?

tallclair replied:
I guess that's fine. As long as the empty runtimeHandler case still defaults to the old behavior, which I think it does.

runcom commented:
@sboeuf could you pick up

https://github.com/kubernetes-sigs/cri-o/pull/1820/commits/ff2c3e1909ec3d1526509bba0a5dd92fc1c9c417
https://github.com/kubernetes-sigs/cri-o/pull/1820/commits/2451ac0a9ae60b1babe4f3f0e10c58f5775e7f78

so this PR becomes our 1.12 PR since you've already cherry-picked the bump PR?
Also, we need to fixup reference to 1.12 under `common/test/...` for stuff like critest and the like

Thanks!

sboeuf commented:
@mrunalp please could you get some feedback about https://github.com/kubernetes-sigs/cri-o/pull/1822#commitcomment-30795678 and https://github.com/kubernetes-sigs/cri-o/pull/1822#commitcomment-30795675 ;)

sboeuf commented:
@runcom @mrunalp  @tallclair 
I have updated this PR:
- Added a new function `ValidateRuntimeHandler()` that is called both from the moment the runtime handler is passed from CRI, and for every time we want to know the runtime path. No duplication of code doing the verification this way.
- I am returning an error if the runtime handler passed by CRI is not valid (`empty` or `not found` or `invalid_value`)
- I have modified the function `(r *runtime) Path(...)` so that it can return an error now. This function is basing the validation of the runtime handler on the same function `ValidateRuntimeHandler()`.
- I have cherry-picked the two commits you mentioned @runcom 
 
I hope this covers all your previous comments, let me know if I missed anything.

runcom commented:
Looks ok, @tallclair would you mind doing a last review on this?

mrunalp commented:
/test all

mrunalp commented:
@sboeuf There are unit test failures in travis:
```
--- FAIL: TestUpdateFromFile (0.00s)
	config_test.go:50: Expecting: "untrusted", got: ""
	config_test.go:50: Expecting: "trusted", got: ""
--- FAIL: TestToFile (0.00s)
	config_test.go:50: Expecting: "untrusted", got: ""
	config_test.go:50: Expecting: "trusted", got: ""
```

sboeuf commented:
@mrunalp I took care of the failure you mentioned

mrunalp commented:
@sboeuf Thanks! It looks like the integration tests are failing in travis.

runcom commented on cmd/crio/config.go:
```diff
@@ -89,7 +89,15 @@ runtime = "{{ .Runtime }}"
 # is an optional setting, except if default_container_trust is set to
 # "untrusted".
-runtime_untrusted_workload = "{{ .RuntimeUntrustedWorkload }}"
+# DEPRECATED: use "crio.runtime.runtimes" instead. If provided, this
```
we should deprecate `Runtime` as well and tell people to use `Runtimes` instead https://github.com/kubernetes-sigs/cri-o/pull/1822/files#diff-c9aa780302e81f9f9029ef37a74bd455R87
Also, we must blank out Runtime and use Runtimes 

I guess I'll make a follow up for that

runcom replied:
Done now with my latest commit

runcom commented on lib/config.go:
```diff
@@ -335,9 +354,8 @@ func DefaultConfig() *Config {
-			Runtime:                  "/usr/bin/runc",
-			RuntimeUntrustedWorkload: "",
-			DefaultWorkloadTrust:     "trusted",
+			Runtime:  "/usr/bin/runc",
```
this should go away and we need just:
```
-                       Runtime:  "/usr/bin/runc",
-                       Runtimes: map[string]oci.RuntimeHandler{},
-
+                       Runtimes: map[string]oci.RuntimeHandler{
+                               "runc": {
+                                       RuntimePath: "/usr/bin/runc"
+                               },
+                       },
```

runcom replied:
basically, Runtime should only be used as fallback for compatibily, otherwise, always prefer RuntimeHandler

runcom commented on lib/container_server.go:
```diff
@@ -129,7 +129,7 @@ func New(ctx context.Context, config *Config) (*ContainerServer, error) {
 	}
 
-	runtime, err := oci.New(config.Runtime, config.RuntimeUntrustedWorkload, config.DefaultWorkloadTrust, config.Conmon, config.ConmonEnv, config.CgroupManager, config.ContainerExitsDir, config.ContainerAttachSocketDir, config.LogSizeMax, config.NoPivot, config.CtrStopTimeout)
+	runtime, err := oci.New(config.Runtime, config.RuntimeUntrustedWorkload, config.DefaultWorkloadTrust, config.Runtimes, config.Conmon, config.ConmonEnv, config.CgroupManager, config.ContainerExitsDir, config.ContainerAttachSocketDir, config.LogSizeMax, config.NoPivot, config.CtrStopTimeout)
```
Runtime and Runtimes here is so confusing 😕

runcom replied:
I'll take care of this in a follow up

runcom commented on oci/oci.go:
```diff
@@ -85,6 +91,7 @@ type Runtime struct {
 	trustedPath              string
 	untrustedPath            string
 	trustLevel               string
+	runtimes                 map[string]RuntimeHandler
```
all this confusion boils down to the fact that a struct named Runtime holds a Runtimes map, omg
I'm just saying we need to fix this :) not in this PR maybe

runcom replied:
I'll take care of this in a follow up

runcom commented on oci/oci.go:
```diff
@@ -112,31 +125,69 @@ func (r *Runtime) Name() string {
 		}
 
-		return r.trustedPath
+		return r.trustedPath, nil
```
trustedPath is basically whatever we have in `Runtime`, I think we need to always prefer Runtimes with a default one

runcom replied:
I believe this is taken care of by the code above this one rigtht? @sboeuf

sboeuf replied:
Yes the code above this will first try to handle a `runtime_handler` if provided.

runcom commented:
@sboeuf I've pushed a commit to this PR to take care of a default for Runtimes and deprecating Runtime

runcom commented:
/test all

runcom commented:
/test all

runcom commented:
/test all

runcom commented:
/test all

runcom commented:
/test all

runcom commented on cmd/crio/config.go:
```diff
@@ -107,7 +115,20 @@ runtime_untrusted_workload = "{{ .RuntimeUntrustedWorkload }}"
+  # The runtime to use is picked based on the runtime_handler provided by the CRI.
+  # If no runtime_handler is provided, the runtime will be picked based on the level
+  # of trust of the workload.
+  [crio.runtime.runtimes.runtime_handler_name]
```
this is killing the configuration parsing in the config subcommand. Basically, every option below this now aren't getting picked up anymore from the config. I've fixed it in the very latest commit

runcom commented:
@sboeuf please rebase this whole PR to latest master as my commits are based on your PR+latest_master

runcom commented:
/test all

runcom commented:
/retest

runcom commented:
tests look busted but not because of this PR

runcom commented:
/retest

sboeuf commented:
@runcom thanks I'll take a look today!

runcom commented:
/retest

runcom commented:
ok, tests are failing because of the merge conflicts :cry: @sboeuf when you rebase and push force, it's going to be working again.

runcom commented:
/test all

runcom commented:
damn, still conflicts after recent merges :)

runcom commented:
Looks like everything is cool in #1847 which is the rebased version of this PR.
I'd say let's rebase this @sboeuf and do another tests run and then we can merge, backport to 1.12 and release 1.12.

Thanks

sboeuf commented:
@runcom I am sorry I couldn't review this PR yesterday, let me do this right now.
About the rebase, why didn't you rebase directly on this one? And that's great to see that #1847 is passing!

runcom commented:
> About the rebase, why didn't you rebase directly on this one? And that's great to see that #1847 is passing!

I haven't cause I didn't want you to git reset --hard because I changed the commits :). We can also close this one and go with #1847, what do you think? that's already rebased and passing and I've cherry-picked your commit from here, sounds good?

sboeuf commented on cmd/crio/main.go:
```diff
@@ -55,6 +55,9 @@ func validateRuntimeConfig(config *server.Config) error {
 	if ok && config.RuntimeConfig.RuntimeUntrustedWorkload != "" {
 		return fmt.Errorf("conflicting definitions: configuration includes runtime_untrusted_workload and runtimes[%q]", oci.UntrustedRuntime)
 	}
+	if config.RuntimeConfig.Runtime != "" {
```
Please move this to the top of this function with the block of `if` statements responsible for checking all the deprecated use cases. A comment would be nice to specify that we're checking all the deprecated behaviors first.

sboeuf commented on oci/oci.go:
```diff
@@ -67,6 +68,15 @@ func New(runtimeTrustedPath string,
+		// the runtime handler configured as default.
+		r, ok := runtimes[defaultRuntime]
+		if !ok {
+			return nil, fmt.Errorf("no runtime configured for default_runtime=%q", defaultRuntime)
```
I feel like we should enforce a check that this default runtime is present from `validateRuntimeConfig()`.

runcom replied:
there's already such check and indeed this is redundant, but still, I want this here for further confirmation

sboeuf commented on cmd/crio/config.go:
```diff
@@ -91,7 +91,7 @@ stream_tls_ca = "{{ .StreamTLSCA }}"
 # default_runtime is the _name_ of the OCI runtime to be used as the default.
 # The name is matched against the runtimes map below.
-default_runtime = {{ .DefaultRuntime }}
+default_runtime = "{{ .DefaultRuntime }}"
```
Please include the fix directly in the commit introducing the change.

runcom replied:
I'll squash them before merging, no worry :)

sboeuf commented on cmd/crio/main.go:
```diff
@@ -58,8 +58,9 @@ func validateRuntimeConfig(config *server.Config) error {
-	if len(config.RuntimeConfig.Runtimes) == 0 && config.RuntimeConfig.Runtime == "" {
-		return errors.New("no deafult runtime configured")
+	_, ok = config.RuntimeConfig.Runtimes[config.RuntimeConfig.DefaultRuntime]
+	if !ok && config.RuntimeConfig.Runtime == "" {
```
This check could be moved above in the code, since I think that's an important check that you want to perform even before you check for any untrusted runtime.
Also, please make sure this is included in the commit introducing `DefaultRuntime`.

runcom replied:
I'll squash them before merging, no worry :)

runcom commented:
@sboeuf if it's ok for you we can go with #1847 , ok to close this?

sboeuf commented:
@runcom oh yeah sure we can use #1847, I'm fine with that, just please take a look at the comments I left on #1822 ;)

runcom commented on cmd/crio/main.go:
```diff
@@ -54,6 +55,9 @@ func validateRuntimeConfig(config *server.Config) error {
 		return fmt.Errorf("conflicting definitions: configuration includes runtime_untrusted_workload and runtimes[%q]", oci.UntrustedRuntime)
 	}
+	if len(config.RuntimeConfig.Runtimes) == 0 && config.RuntimeConfig.Runtime == "" {
+		return errors.New("no deafult runtime configured")
```
Errorf is with string formatting, I don't need C-like string formatting here, hence the use of the golang errors package. This is pretty common afaict

runcom replied:
fixed typo

runcom commented on cmd/crio/config.go:
```diff
@@ -127,8 +130,10 @@ runtime = "{{ .Runtime }}"
   # of trust of the workload.
-  [crio.runtime.runtimes.runtime_handler_name]
-    runtime_path = ""
+  {{ range $runtime_name, $runtime_path := .Runtimes  }}
```
agree, lol, I hadn't paid that much attention to naming cause i know i suck :P

runcom commented:
> @runcom oh yeah sure we can use #1847, I'm fine with that, just please take a look at the comments I left on #1822 ;)

awesome, thanks a lot for the comments @sboeuf I'll fix them in #1847 and ping you there once ready for another, hopefully final, review

sboeuf commented:
@runcom 
>awesome, thanks a lot for the comments @sboeuf I'll fix them in #1847 and ping you there once ready for another, hopefully final, review

Cool :smiley:

runcom commented on cmd/crio/main.go:
```diff
@@ -54,6 +55,9 @@ func validateRuntimeConfig(config *server.Config) error {
 	if ok && config.RuntimeConfig.RuntimeUntrustedWorkload != "" {
 		return fmt.Errorf("conflicting definitions: configuration includes runtime_untrusted_workload and runtimes[%q]", oci.UntrustedRuntime)
 	}
+	if len(config.RuntimeConfig.Runtimes) == 0 && config.RuntimeConfig.Runtime == "" {
```
done

runcom commented:
taken care of all your comments in #1847 , let's move the discussion there now @sboeuf
